### PR TITLE
CSHARP-3173: CSharp driver does not fallback to a healthy mongos after one is unavailable.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
EG: https://evergreen.mongodb.com/version/5f21cb8ae3c3311bc5cbab2f

### CODE REVIEW

**Review by rstam — APPROVED** [review]
LGTM

Note: the CSharp3173Tests.cs could use a LOT of refactoring to improve it but in the interest of time and at Jeff's suggestion we won't spend any time reviewing this test.

File: src/MongoDB.Driver.Core/Core/Servers/Server.cs [hunk 1]
```diff
@@ -331,6 +331,13 @@ private void HandleBeforeHandshakeCompletesException(IConnection connection, Exc
                     return; // stale generation number
                 }
 
+                if (ex is MongoConnectionException mongoConnectionException &&
+                    mongoConnectionException.IsNetworkException &&
+                    !mongoConnectionException.ContainsSocketTimeoutException)
```
Comment (jyemin) [inline src/MongoDB.Driver.Core/Core/Servers/Server.cs:336]: Can a connection exception also be a socket time exception?  And if so: if connect times out, would we not still want to cancel the current check?

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs [hunk 2]
```diff
@@ -378,7 +378,11 @@ private async Task HeartbeatAsync(CancellationToken cancellationToken)
 
                 newDescription = newDescription.With(reasonChanged: "Heartbeat", lastHeartbeatTimestamp: DateTime.UtcNow);
 
-                SetDescription(newDescription);
+                lock (_lock)
```
Comment (jyemin) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:381]: Why was this change required?

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:381]: In case if we cancel the current isMaster check after these lines: https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/9e90b5b90479cc69f1bb78d550e0ac5e62c6733e/src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs#L329-L332 but before publishing it. In this case, without this line we will still publish the cancelled isMaster data

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:381]: Maybe a better option, move lines 329-332 to line 383 for consistency

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:381]: NOTE: in the current implementation, the thrown exception will be caught here https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/9e90b5b90479cc69f1bb78d550e0ac5e62c6733e/src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs#L211-L214

File: src/MongoDB.Driver.Core/Core/Servers/Server.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/Server.cs:336]: >Can a connection exception also be a socket time exception?

If you ask about how it's implemented in the C#, then yes. MongoConnectionException  contains different cases because actually this exception wraps a number of cases from BinaryConnection https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L725-L744

As far as I know, ContainsSocketTimeoutException is the connection timeout exception with the code: `SocketError.TimedOut`. The spec says that we need to cancel only `Network error`.

This is a note: `When an application operation fails with a non-timeout network error, drivers cancel that monitor's in-progress check.` from here:https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-monitoring.rst#drivers-cancel-in-progress-monitor-checks

It also mimics the same logic in AFTER handshake error processing. Do you think we need to reconsider this logic?

File: tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        private readonly static ClusterId __clusterId = new ClusterId();
+        private readonly static EndPoint __endPoint1 = new DnsEndPoint("localhost", 27017);
+        private readonly static EndPoint __endPoint2 = new DnsEndPoint("localhost", 27018);
+        private readonly static TimeSpan __heartbeatInterval = TimeSpan.FromMilliseconds(200);
+        private readonly static ServerId __serverId1 = new ServerId(__clusterId, __endPoint1);
+        private readonly static ServerId __serverId2 = new ServerId(__clusterId, __endPoint2);
+
+        [Theory]
+        [ParameterAttributeData]
+        public void Ensure_command_network_error_before_hadnshake_is_correctly_handled([Values(false, true)] bool async, [Values(false, true)] bool streamable)
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs:52]: Logic:
1. we launch heartbeat
2. we select the server (27017)
3. Then we pause the isMaster response for this server until the GetChannel will trigger network exception
4. Then we cancel the current isMaster paused check. After that we come to this line: https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/6ed472e10c07b86f098691ac7ad80dd7e91296e7/src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs#L331 So, the canceled description is not published

NOTE: This case is not covered by the test since I didn't find the way to mock it, but it's also possible that we canceled the master check after line 331. In this case, we will throw cancellation exception in this line: https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/6ed472e10c07b86f098691ac7ad80dd7e91296e7/src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs#L383

File: tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+            // ensure that isMaster check response is finished only after network error
+            var hasNetworkErrorBeenTriggered = new TaskCompletionSource<bool>();
+            // ensure that there are no unexpected events between test ending and cluster disposing
+            var hasClusterBeenDisposed = new TaskCompletionSource<bool>();
+
+            EndPoint initialSelectedEndpoint = null;
+            using (var cluster = CreateAndSetupCluster(hasNetworkErrorBeenTriggered, hasClusterBeenDisposed, eventCapturer, streamable))
+            {
+                cluster._clusterId(__clusterId);
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs]: I don't know how you can get away with this. The cluster implementation assumes that the ClusterId never changes.

One symptom of how this isn't really a good idea... during the whole test the ClusterId in the ClusterDescription does NOT match the ClusterId that you forcibly overwrote on this line.

I'm not sure why these mismatched ClusterIds don't cause problems, but it concerns me that we are forcibly overwriting the ClusterId value when the class is not expecting that.

In fact, the `_clusterId` field is declared **readonly** which only further emphasizes that it should not be changed.

**Review by rstam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs]: I'm not sure that the whole question is about the scope of this particular test (from my point of view, the target of the test is clearly defined and pretty specific) :) 

>The cluster implementation assumes that the ClusterId never changes.

I don't think that it's significantly important for emulating the particular test scenario. And anyway the clusterId is never changed **after** the time when the cluster has been initialized.

>One symptom of how this isn't really a good idea... during the whole test the ClusterId in the ClusterDescription does NOT match the ClusterId that you forcibly overwrote on this line.

I'm not sure that it's important for the scope of the current test, clusterId in ClusterDescription is used only in events. However, it's fixed in the latest commit.

>I'm not sure why these mismatched ClusterIds don't cause problems

most of the clusterId usage is about sdam logging.

>In fact, the _clusterId field is declared readonly which only further emphasizes that it should not be changed.

I don't think that this restriction is about mocking in the test. In fact, all of our reflection usages is a hack to avoid some code design restrictions.

**Comment by DmitryLukyanov** [comment]
ready for review

**Comment by DmitryLukyanov** [comment]
EG: https://evergreen.mongodb.com/version/5f25f4412a60ed7366550865

File: tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs (hunk 4 above)
Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs]: I wouldn't call using reflection in tests a hack... the alternative is to declare things that you normally want to be private internal or worse yet public, and that would be more like a hack.

I still think overwriting the `_clusterId` is not a good technique... and overwriting the `_description` also to compensate doesn't make it any better.

I understand why you are doing this... but I don't think it's the only approach or the best one. But in the interest of expediency I guess I'm willing to go along with this. I could try to find a better approach but not sure I have the time.

Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs]: >I wouldn't call using reflection in tests a hack...

it's a hack, in meaning that we break the design restriction. But it's not a hack in meaning that these restrictions are not created for test code or at very least they are not significantly important for the test code.

>but I don't think it's the only approach or the best one. 

I'm sure it's possible to do it in another way too, but I don't think that this makes any sense for the scope/target of this test :)

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs [hunk 5]
```diff
@@ -614,6 +614,8 @@ private void AddOrRemoveServers(ClusterDescription clusterDescription)
 
     internal static class ClusterReflector
     {
+        public static void _clusterId(this Cluster cluster, ClusterId clusterId) => Reflector.SetFieldValue(cluster, nameof(_clusterId), clusterId);
+        public static void _description(this Cluster cluster, ClusterDescription clusterDescription) => Reflector.SetFieldValue(cluster, nameof(_description), clusterDescription);
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs]: Remove these two lines from `ClusterReflector`. Instead put this logic in a helper method the CSharp3173Tets class:

    private void ForceClusterId(MultiServerCluster cluster, ClusterId clusterId)
    {
        Reflector.SetFieldValue(cluster, "_clusterId", clusterId);
        Reflector.SetFieldValue(cluster, "_description", ClusterDescription.CreateInitial(__clusterId, __clusterConnectionMode));
    }


I don't want to encourage any other tests from overwriting these fields.

File: tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs (hunk 4 above)
Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs]: Replace 63-64 with:

    ForceClusterId(cluster, __clusterId);

And add new helper method in THIS class:

    private void ForceClusterId(MultiServerCluster cluster, ClusterId clusterId)
    {
        Reflector.SetFieldValue(cluster, "_clusterId", clusterId);
        Reflector.SetFieldValue(cluster, "_description", ClusterDescription.CreateInitial(__clusterId, __clusterConnectionMode));
    }

**Review by rstam — CHANGES_REQUESTED** [review]

**Comment by rstam** [comment]
Comment: I don't really understand how the test works in detail. At a high level I understand that it sets up the scenario to be tested, but at a low level there are two many details and steps to review in a time efficient manner.

I'm just going to wave my hands and say since the test passes it must be OK (and acknowledge that while I don't agree with the technique of overwriting the _clusterId that it can be left that way in the interest of time).

File: src/MongoDB.Driver.Core/Core/Servers/Server.cs (hunk 1 above)
Comment (jyemin) [reply src/MongoDB.Driver.Core/Core/Servers/Server.cs:336]: I don't really know.  I will defer to you and @rstam

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs (hunk 2 above)
Comment (jyemin) [reply src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:381]: I will defer to you and @rstam

**Comment by DmitryLukyanov** [comment]
EG:  https://evergreen.mongodb.com/version/5f28572f9ccd4e199193e40c

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs]: done

File: tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs]: done

**Comment by jyemin** [comment]
@rstam  can you have a look at my comment threads and decide if there's anything further that needs to be done?

File: src/MongoDB.Driver.Core/Core/Servers/Server.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Servers/Server.cs:336]: @dima what exception is thrown if a connection times out? Do we need to tell the difference between a timeout in the initial connection attempt vs a timeout in a subsequent I/O operation (like isMaster)? If we need to tell the difference can we?

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs (hunk 2 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:381]: I'm fine with adding this lock statement.

**Review by rstam — COMMENTED** [review]
Waiting for a comment thread to be resolved before approving.

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:381]: I think now the main question here is moving lines https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/9e90b5b90479cc69f1bb78d550e0ac5e62c6733e/src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs#L329-L332 to 383 (instead of throwing exception). 
Actually there is no real difference where we will leave the current method. And after applying of the proposed change, we will have exactly the same behavior as we have now. And additionally, we won't have a new different way to leave the method via throwing cancellation exception

File: src/MongoDB.Driver.Core/Core/Servers/Server.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/Server.cs:336]: >@dima what exception is thrown if a connection times out?

it looks like we can have two cases:
1. If a custom driver timeout exception is thrown from here:  https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Connections/TcpStreamFactory.cs#L158. In this case, the underlying exception will be `TimeoutException`:

    >e.GetType().Name
    "MongoConnectionException"
    >e.Message
    "An exception occurred while opening a connection to the server."
    >e.InnerException.GetType().Name
    "TimeoutException"
    >e.InnerException.Message
    "Timed out connecting to [::1]:27017. Timeout was 00:00:00." // I set ConnectTimeout=0 to emulate this case
    >e.InnerException.InnerException
    null

At first glance, this case is not covered by our logic and we should create a new ticket to that
2. The flag which was added in the scope of reworking CSHARP-2126: https://github.com/mongodb/mongo-csharp-driver/commit/5caa4eeaf30549b0f531761cdb496aced98104bc#diff-02a6d9e62d0d60a22a25f4f5d8178693R86. This flag will be set when socket itself triggers the timeout exception.
We definitely discussed the first case with Vincent, but I don't remember why it wasn't added in the code (maybe there are some side effects for that)

>Do we need to tell the difference between a timeout in the initial connection attempt vs a timeout in a subsequent I/O operation

I don't see it in the spec. So, probably no:) The spec says:
     
     When a client marks a server Unknown from Network error when reading or writing

and I think it's clearly said here: https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-monitoring.rst#drivers-cancel-in-progress-monitor-checks

>When an application operation fails with a ***non-timeout network error***, drivers cancel that monitor's in-progress check

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/Server.cs:336]: In two words, I would say, no changes are required right now, but we need to investigate why the TimeoutException from the driver side was not considered for this logic (I think it will also require changes here: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Servers/Server.cs#L301)

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/Server.cs:336]: Vincent's comment about ContainsSocketTimeoutException: https://github.com/vincentkam/mongo-csharp-driver/pull/114#discussion_r428173287
So, I think both of the above points should be considered as Timeout exception. But I still think it's better to change it in the scope of a new ticket

**Review by rstam — APPROVED** [review]
LGTM

I have no pending comments. @DmitryLukyanov did you need anything from me?

**Comment by DmitryLukyanov** [comment]
>did you need anything from me?

these ones :) :
1. https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/140#discussion_r467337830. The suggestion about consistent handling cancellation exceptions in two different places (the idea is just having only old one but in line 383 )
2. https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/140#discussion_r469266325 it looks like we missed previously one case for timeout handling, so it should be investigated additionally in a new ticket I guess

**Comment by rstam** [comment]
> > did you need anything from me?
> 
> these ones :) :
> 
> 1. [#140 (comment)](https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/140#discussion_r467337830). The suggestion about consistent handling cancellation exceptions in two different places (the idea is just having only old one but in line 383 )
> 2. [#140 (comment)](https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/140#discussion_r469266325) it looks like we missed previously one case for timeout handling, so it should be investigated additionally in a new ticket I guess

1. I have already commented that the normal response to cancellation requested is to throw an OperationCanceledException. But I would not require you to change existing code.

2. OK create a new ticket.
